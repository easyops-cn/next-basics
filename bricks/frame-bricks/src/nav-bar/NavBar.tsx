/* eslint-disable @typescript-eslint/no-use-before-define */
import React, { useState } from "react";
import { Menu } from "antd";
import {
  SidebarMenuSimpleItem,
  SidebarMenuItem,
  SidebarMenuGroup,
} from "@next-core/brick-types";
import classNames from "classnames";
import { GeneralIcon, Link } from "@next-libs/basic-components";
import style from "./NavBar.module.css";

interface SidebarMenuProps {
  menuItems: SidebarMenuItem[];
  selectedKeys: string[];
}

function isGroup(item: SidebarMenuItem): item is SidebarMenuGroup {
  return item.type === "group";
}

function isSubMenu(
  item: SidebarMenuItem,
  showMenu?: boolean
): item is SidebarMenuGroup {
  return item.type === "subMenu" || (showMenu && item.type === "group");
}

export function NavBar(props: SidebarMenuProps): React.ReactElement {
  const { menuItems, selectedKeys } = props;
  const [selectedKey, setSelectedKey] = useState(selectedKeys ?? []);

  const renderSimpleMenuItem = (
    item: SidebarMenuSimpleItem,
    showEmptyIcon?: boolean
  ): React.ReactNode => {
    return (
      <Menu.Item
        key={String(item.key)}
        title={item.text}
        className={style.simpleMenuItem}
      >
        <Link to={item.to} href={item.href} target={item.target}>
          {showEmptyIcon ? (
            <i className={style.menuItemIcon}></i>
          ) : (
            item.icon && (
              <i className={style.menuItemIcon}>
                <GeneralIcon icon={item.icon} size={14} />
              </i>
            )
          )}
          <span
            className={classNames(style.menuText, style.simpleMenuItemText)}
          >
            {item.text}
          </span>
        </Link>
      </Menu.Item>
    );
  };

  const renderGroupMenu = (
    item: SidebarMenuGroup,
    showEmptyIcon?: boolean
  ): React.ReactNode => {
    return (
      <Menu.ItemGroup
        key={item.key}
        className={style.groupWrapper}
        title={
          <span>
            <i
              className={classNames(style.menuItemIcon, style.groupTitlePoint, {
                [style.hideGroupTitlePoint]: showEmptyIcon,
              })}
            ></i>
            <span className={classNames(style.menuText, style.groupText)}>
              {item.title}
            </span>
          </span>
        }
      >
        {item.items?.map((innerItem) =>
          renderMenuItem(innerItem, showEmptyIcon)
        )}
      </Menu.ItemGroup>
    );
  };

  const renderSubMenu = (item: SidebarMenuGroup): React.ReactNode => {
    return (
      <Menu.SubMenu
        key={item.key}
        className={style.subMenuWrapper}
        popupClassName={style.popupWrapper}
        title={
          <span>
            {item.icon && (
              <i className={style.menuItemIcon}>
                <GeneralIcon icon={item.icon} size={14} />
              </i>
            )}
            <span
              className={classNames(style.menuText, style.subMenuTitleText)}
            >
              {item.title}
            </span>
          </span>
        }
      >
        {item.items?.map((innerItem) => renderMenuItem(innerItem, true))}
      </Menu.SubMenu>
    );
  };

  const renderMenuItem = (
    item: SidebarMenuItem,
    showEmptyIcon?: boolean,
    showSubMenu?: boolean
  ): React.ReactNode => {
    // return renderGroupMenu(item, showEmptyIcon)
    return isSubMenu(item, showSubMenu)
      ? renderSubMenu(item)
      : isGroup(item)
      ? renderGroupMenu(item, showEmptyIcon)
      : renderSimpleMenuItem(item, showEmptyIcon);
  };

  return (
    <Menu
      mode="horizontal"
      selectedKeys={selectedKey}
      className={style.navBarContainer}
      onClick={(e) => setSelectedKey([e.key as string])}
    >
      {menuItems.map((item) => renderMenuItem(item, undefined, true))}
    </Menu>
  );
}
