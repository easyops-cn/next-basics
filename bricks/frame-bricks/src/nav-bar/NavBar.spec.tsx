import React from "react";
import { Menu } from "antd";
import { mount } from "enzyme";
import { NavBar } from "./NavBar";
import { SidebarSubMenu } from "@next-core/brick-types";
import { act } from "react-dom/test-utils";

const mockData = {
  title: "二级菜单标题",
  link: "/nlicro-test/aaaa",
  menuItems: [
    {
      type: "subMenu",
      title: "subMenu",
      icon: {
        color: "geekblue",
        icon: "anchor",
        lib: "fa",
        prefix: "fas",
      },
      items: [
        {
          type: "group",
          title: "group-1",
          icon: {
            color: "geekblue",
            icon: "anchor",
            lib: "fa",
            prefix: "fas",
          },
          items: [
            {
              text: "subMenu-group-item-1",
              type: "default",
              key: "1",
              icon: {
                color: "geekblue",
                icon: "anchor",
                lib: "fa",
                prefix: "fas",
              },
            },
            {
              text: "subMenu-group-item-2",
              type: "default",
              key: "2",
              icon: {
                color: "geekblue",
                icon: "anchor",
                lib: "fa",
                prefix: "fas",
              },
            },
          ],
        },
      ],
    },
    {
      type: "group",
      title: "group-2",
      items: [
        {
          type: "default",
          text: "group-2-item-1",
          key: "3",
        },
      ],
    },
    {
      type: "default",
      text: "page-3",
      key: "4",
    },
  ],
} as SidebarSubMenu;

describe("NavBar", () => {
  it("should work", () => {
    const wrapper = mount(
      <NavBar menuItems={mockData.menuItems} selectedKeys={["3"]} />
    );

    expect(wrapper.find(Menu).prop("selectedKeys")).toEqual(["3"]);

    expect(wrapper.find(Menu.SubMenu).length).toBe(2);

    expect(wrapper.find(Menu.Item).length).toBe(1);

    wrapper.find(Menu.Item).last().simulate("click");

    wrapper.update();

    expect(wrapper.find(Menu).prop("selectedKeys")).toEqual(["4"]);
  });
});
