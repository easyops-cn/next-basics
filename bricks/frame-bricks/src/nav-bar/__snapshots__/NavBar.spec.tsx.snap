// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`NavBar should work 1`] = `
<div
  className="nav-bar"
>
  <div
    className="left-content"
  >
    <div
      className="nav-logo"
    >
      <slot
        id="logo"
        name="logo"
      />
    </div>
    <div
      className="nav-menu"
    >
      <slot
        id="menu"
        name="menu"
      />
    </div>
  </div>
  <div
    className="right-content"
  >
    <div
      className="nav-buttons"
    >
      <slot
        id="buttons"
        name="buttons"
      />
    </div>
    <div
      className="nav-avator"
    >
      <slot
        id="avator"
        name="avator"
      />
    </div>
  </div>
</div>
`;
